### ISSUE TITLE
Trainer wraps the weights of the model in extra initialized classes.

### ISSUE THREAD

**Issue by sinAshish** [Lightning-AI/lightning#18066]
### Bug description

I have a PL model. which is defined as:
```python
class PL_Model(pl.LightningModule):
    def __init__(self, opt):
        super().__init__() 
        self.opt = opt
        self.model = AwesomeModel(self.opt)
        self.prior_function= ComputePrior(config = self.opt)
  def forward(...):
....#usual PL logic
```
I am using ModelCheckpoint callback to monitor the `val/avg_iou` metric to save the entire checkpoint and not just the weights.

However, while loading the checkpoints for testing, in a seperate script( not after calling `.fit()`), it loads the keys in a nested manner, such as:

```python
{"self.model.{conv/fc/ blah blah blah}, 
self.prior_function.model.{conv/fc/ blah blah blah}, }
```
the same thing but this time the model is nested within another unnecessary key, which doesn't have any learnable parameters.
This throws an error while loading the model checkpoints.
I can set `strict=False` to solve this, but I don't understand why this should happen.
Even though, ComputePrior is `nn.Module` class, it should have its parameters sepearate, or everything should be nested inside 

### What version are you seeing the problem on?

v1.7

### Error messages and logs

```
Missing keys
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0): 1.7.5
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0): 1.12.1
#- Python version (e.g., 3.9): 3.8
#- OS (e.g., Linux):
#- CUDA/cuDNN version: 11.3
#- GPU models and configuration: 1
#- How you installed Lightning(`conda`, `pip`, source): pip
#- Running environment of LightningApp (e.g. local, cloud): local
```

</details>

cc @awaelchli

**Comment by awaelchli** [comment]
@sinAshish I'm sorry I don't quite understand your explanation. How did you end up with this?

```
{"self.model.{conv/fc/ blah blah blah}, 
self.prior_function.model.{conv/fc/ blah blah blah}, }
```

This doesn't look anything like a state dict of a model. Would you mind please sharing a concise, runnable code example?

FYI, in PyTorch, it is normal that if you have a module nested under another module, you get the weights with a dotted name like "module.submodule.subsubmodule.weight" etc. This is indeed how it is supposed to be.

**Comment by sinAshish** [comment]
the keys should be like:
`"<lightning_model_name>.<module1>.submodule.weight"`

however, in my case:
apart from `"<lightning_model_name>.<module1>.submodule.weight"`, there are other keys:
`"<lightning_model_name>.<module2>.<module1>.submodule.weight"`

even though `module2` and `moduleq` are both initialised in the `___init__()` of PL module.

**Comment by awaelchli** [comment]
@sinAshish I'm sorry, but I still don't understand what you are describing. This does not at all look like a Lightning checkpoint. Maybe you have implemented some custom checkpointing? Please explain the problem you are facing with an accurate description and an example code that runs, or I'll will close the issue due to insufficient information.

Here is a code snippet that demonstrates how the keys are named in the checkpoint file:

```py
import torch
from torch.utils.data import DataLoader, Dataset

from lightning.pytorch import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.model = torch.nn.Linear(32, 2)
        self.prior_function = torch.nn.Sequential(torch.nn.Linear(5, 5), torch.nn.Linear(5, 2))

    def forward(self, x):
        return self.model(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(max_steps=1)
    trainer.fit(model, train_dataloaders=train_data)

    checkpoint = torch.load(trainer.checkpoint_callback.best_model_path)
    print(list(checkpoint["state_dict"].keys()))


if __name__ == "__main__":
    run()

```

This prints:

```
['model.weight', 'model.bias', 'prior_function.0.weight', 'prior_function.0.bias', 'prior_function.1.weight', 'prior_function.1.bias']
```
As you can see, the state-dict contains the dotted module names of all submodules, named after their attributes in the LightningModule. Please take this example code and modify it to your use case to show what the issue is you are facing. Only if you provide us with an accurate description are we able to help, or otherwise we are just spending valuable time guessing the unknown. Thank you!

**Comment by awaelchli** [comment]
Hey @sinAshish 
Did my answer help, or were you able to resolve the problem?

**Comment by sinAshish** [comment]
@awaelchli 
i understand your answer, and that's how the model keys are strored. 
since I am a bit busy with a  deadline, i'll provide the model checkpoint and a minimal example in the format your suggested in a few days.

**Comment by sinAshish** [comment]
meanwhile, I just turn off strict matching of keys to get past this issue!

**Comment by awaelchli** [comment]
I'll close the ticket now, but if we find evidence that there is an issue with how Lightning is handling this (through a reproducible example) we can certainly reopen and look into it again.
